pkg/controller/walle/controller.go

```diff
@@ -0,0 +1,435 @@
+	DeleteFunc func(kluster *v1.Kluster, node *api_v1.Node)
+
+	WalleController struct {
+		Controller
+
+		config.Factories
+		config.Clients
+		queue               workqueue.RateLimitingInterface
+		logger              log.Logger
+		nodeInformerMap     sync.Map
+		handlersMux         sync.RWMutex
+		addEventHandlers    []AddFunc
+		updateEventHandlers []UpdateFunc
+		deleteEventHandlers []DeleteFunc
+		stopCh              <-chan struct{}
+		informerWg          *sync.WaitGroup
+	}
+)
+
+func NewController(factories config.Factories, clients config.Clients, logger log.Logger) *WalleController {
+	logger = log.With(logger,
+		"controller", "walle",
+	)
+
+	controller := &WalleController{
+		Factories:       factories,
+		Clients:         clients,
+		queue:           workqueue.NewRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(BaseDelay, MaxDelay)),
+		logger:          logger,
+		nodeInformerMap: sync.Map{},
+	}
+
+	controller.Factories.Kubernikus.Kubernikus().V1().Klusters().Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    controller.klusterAddFunc,
+		UpdateFunc: controller.klusterUpdateFunc,
+		DeleteFunc: controller.klusterDeleteFunc,
+	})
+
+	return controller
+}
+
+func (wr *WalleController) Run(threadiness int, stopCh <-chan struct{}, wg *sync.WaitGroup) {
+	wr.logger.Log(
+		"msg", "starting run loop",
+		"threadiness", threadiness,
+		"v", 2,
+	)
+
+	wr.stopCh = stopCh
+	wr.informerWg = wg
+
+	defer wr.queue.ShutDown()
+	defer wg.Done()
+	wg.Add(1)
+
+	for i := 0; i < threadiness; i++ {
+		go wait.Until(wr.runWorker, time.Second, stopCh)
+	}
+
+	ticker := time.NewTicker(KlusterRecheckInterval)
+	go func() {
+		for {
+			select {
+			case <-ticker.C:
+				wr.requeueAllKlusters()
+			case <-stopCh:
+				ticker.Stop()
+				return
+			}
+		}
+	}()
+	<-stopCh
+}
+
+func (wr *WalleController) requeueAllKlusters() (err error) {
+	defer func() {
+		wr.logger.Log(
+			"msg", "requeued all",
+			"v", 1,
+			"err", err,
+		)
+	}()
+
+	klusters, err := wr.Factories.Kubernikus.Kubernikus().V1().Klusters().Lister().List(labels.Everything())
+	if err != nil {
+		return err
+	}
+
+	for _, kluster := range klusters {
+		wr.requeueKluster(kluster)
+	}
+
+	return nil
+}
+
+func (wr *WalleController) requeueKluster(kluster *v1.Kluster) {
+	wr.logger.Log(
+		"msg", "queuing",
+		"kluster", kluster.Spec.Name,
+		"project", kluster.Account(),
+		"v", 2,
+	)
+
+	key, err := cache.MetaNamespaceKeyFunc(kluster)
+	if err == nil {
+		wr.queue.Add(key)
+	}
+}
+
+func (wr *WalleController) runWorker() {
+	for wr.processNextWorkItem() {
+	}
+}
+
+func (wr *WalleController) processNextWorkItem() bool {
+	key, quit := wr.queue.Get()
+	if quit {
+		return false
+	}
+	defer wr.queue.Done(key)
+
+	var err error
+	var kluster *v1.Kluster
+	var requeue bool
+
+	obj, exists, _ := wr.Factories.Kubernikus.Kubernikus().V1().Klusters().Informer().GetIndexer().GetByKey(key.(string))
+	if !exists {
```

Comment:
I think this is logic is problematic as `reconcile` is not invoked when a kluster delete event is processed. In that case the `exists` will be false. Eventually the `KlusterRecheckInterval` will take care of it but its suboptimal. I think this method should handle the deleted case better.

Maybe its better to defer getting the `Kluster` from the cache to the `reconcile` method and just pass the `key`, or we just pass `nil` to reconcile and that means *something was deleted please resync your local state accordingly*.

Reply:
Pass `nil` to reconcile, in which we invoke `cleanUpInformers`.  Here we'll compare the internal nodeInformerMap with the store. If there's something in the map, which cannot be found in the store, we'll stop and delete it.